sound/soc/sof/intel/hda.c

```diff
@@ -358,18 +358,61 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 
 static int hda_init_caps(struct snd_sof_dev *sdev)
 {
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
```

Comment:
Sorry, I don't get what you are trying to fix.

there are two versions on hda_init_caps, and you are moving common code in _probe() into the second version of hda_init_caps. That makes me wonder if this really works.
Also you didn't provide any details on how the registers are set twice so it's difficult to figure out the problem statement.

Reply:
The commit message is also unclear, what is this none HDA thing you are referring to?

Reply:
> there are two versions on hda_init_caps, and you are moving common code in _probe() into the second version of hda_init_caps. That makes me wonder if this really works.

I have tested this on Up2 with nocodec mode, HDMI mode and IIS mode, this works. and I will go on test with more platforms.

Reply:
> Also you didn't provide any details on how the registers are set twice so it's difficult to figure out the problem statement.

The call stacks are as below:

```
hda_dsp_probe=>
  ......
  hda_init_caps=>
#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
      hda_codec_i915_init
      hda_dsp_ctrl_init_chip=>
            hda_dsp_ctrl_misc_clock_gating
            snd_hdac_bus_init_chip=>
```
**we have done the registers read/write here with HDA lib as follows:**
```
                  snd_hdac_bus_reset_link=>
                        /* clear STATESTS */
                        snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
                        /* reset controller */
                        snd_hdac_bus_enter_link_reset(bus);
                        /* Bring controller out of reset */
                        snd_hdac_bus_exit_link_reset(bus);
                        ......
                  azx_int_clear=>
                        /* clear stream status */
	                list_for_each_entry(azx_dev, &bus->stream_list, list)
		                snd_hdac_stream_writeb(azx_dev, SD_STS, SD_INT_MASK);
	                /* clear STATESTS */
	                snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
	                /* clear rirb status */
	                snd_hdac_chip_writeb(bus, RIRBSTS, RIRB_INT_MASK);
	                /* clear int status */
	                snd_hdac_chip_writel(bus, INTSTS, ...);
                  snd_hdac_bus_init_cmd_io
                  azx_int_enable
```   

**till here**

```
         hda_dsp_ctrl_misc_clock_gating
      ......
      hda_codec_probe_bus
      hda_codec_i915_put
#else
      hda_dsp_ctrl_misc_clock_gating
      snd_sof_dsp_update_bits(SOF_HDA_WAKESTS)
#endif
```
**and we do it here again as common code:**
```
      hda_dsp_ctrl_link_reset(true)   //same as snd_hdac_bus_enter_link_reset(bus);
      hda_dsp_ctrl_link_reset(false)  //same as snd_hdac_bus_exit_link_reset(bus);
      /* clear stream status */          //same as snd_hdac_stream_writeb(SD_STS)
      list_for_each_entry(stream, &bus->stream_list, list) {
            snd_sof_dsp_update_bits(SOF_HDA_ADSP_REG_CL_SD_STS)
      }
      /* clear WAKESTS */                //same as snd_hdac_chip_writew(...STATESTS...);
      snd_sof_dsp_update_bits(SOF_HDA_WAKESTS)
      /* clear interrupt status register */ //same as snd_hdac_chip_writel(...INTSTS ...);
      snd_sof_dsp_write(SOF_HDA_INTSTS)
      /* enable CIE and GIE interrupts */   //same as azx_int_enable
      snd_sof_dsp_update_bits(SOF_HDA_INTCTL)

      hda_dsp_ctrl_misc_clock_gating(sdev, true);
      device_disable_async_suspend(&pci->dev);

```
So if **IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)** is true, then the sequences
```
/* link reset */ 
/* clear stream status */ 
/* clear WAKESTS */ 
/* clear interrupt status register */ 
/* enable CIE and GIE interrupts */ 
```
has been done twice.

Reply:
> The commit message is also unclear, what is this none HDA thing you are referring to?

I mean when **IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)** is **false**. Place where the hda lib is not used.

Reply:
> and I will go on test with more platforms.

I have test on WHL with HDA generic machine driver, this also works.

Reply:
so your commit message is completely misleading. You are not "refining" the code for the case where SOF_HDA is not defined, you are correcting the case where SOF_HDA is defined and moving the code outside of this conditionally-compiled code so that the registers are only initialized once.
Also please describe in the commit message (or PR header) what you wrote in the comments, it's very helpful for reviewers to figure out what steps you considered to double-check the logic. Thanks!

Reply:
@plbossart OK, thanks. Yes, this is actually a "correcting". Will update.